add length assertion to PrepareModuleInput and PrepareModuleOutput

tianyu-l opened the pull request:
## summary

`zip(inputs, self.input_layouts, self.desired_input_layouts)` is used in `_prepare_input_fn`; similar for `_prepare_output_fn`. Without assertion, unmatched dimension in inputs/outputs will be lost, potentially causing unexpected behabiors.

## test plan
`python test/distributed/tensor/parallel/test_tp_style.py`


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115957



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

tianyu-l committed:
add length assertion to PrepareModuleInput and PrepareModuleOutput

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, could you add some tests to test the assertion in test_style.py?

tianyu-l committed:
Update on "add length assertion to PrepareModuleInput and PrepareModuleOutput"

## summary

`zip(inputs, self.input_layouts, self.desired_input_layouts)` is used in `_prepare_input_fn`; similar for `_prepare_output_fn`. Without assertion, unmatched dimension in inputs/outputs will be lost, potentially causing unexpected behabiors.

## test plan
`python test/distributed/tensor/parallel/test_tp_style.py`
`python test/distributed/_tensor/test_dtensor_compile.py`





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l commented:
> lgtm, could you add some tests to test the assertion in test_style.py?

Makes sense! Added some tests to existing PrepareModuleInput test `test_prepare_module_input_multiple_inputs`. Didn't test the PrepareModuleOutput counterpart since functionality test for that is also missing. Let's add later if needed.

tianyu-l commented:
@pytorchbot merge

tianyu-l commented:
@pytorchbot merge
